src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala

```diff
@@ -0,0 +1,140 @@
+
+import com.microsoft.hyperspace.{Hyperspace, HyperspaceException, SampleData}
+import com.microsoft.hyperspace.util.FileUtils
+
+/**
+ * Unit E2E test cases for RefreshIndex.
+ */
+class RefreshIndexTests extends HyperspaceSuite with SQLHelper {
+  override val systemPath = new Path("src/test/resources/indexLocation")
+  private val testDir = "src/test/resources/RefreshIndexDeleteTests/"
+  private val nonPartitionedDataPath = testDir + "nonpartitioned"
+  private val partitionedDataPath = testDir + "partitioned"
+  private val indexConfig = IndexConfig("index1", Seq("Query"), Seq("imprs"))
+  private var hyperspace: Hyperspace = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    hyperspace = new Hyperspace(spark)
+    FileUtils.delete(new Path(testDir))
+  }
+
+  override def afterAll(): Unit = {
+    FileUtils.delete(new Path(testDir))
+    super.afterAll()
+  }
+
+  after {
+    FileUtils.delete(new Path(testDir))
+    FileUtils.delete(systemPath)
+  }
+
+  test("Validate refresh index when some file gets deleted from the source data.") {
+    // save test data non-partitioned.
+    SampleData.save(
+      spark,
+      nonPartitionedDataPath,
+      Seq("Date", "RGUID", "Query", "imprs", "clicks"))
+    val nonPartitionedDataDF = spark.read.parquet(nonPartitionedDataPath)
+
+    // save test data partitioned.
+    SampleData.save(
+      spark,
+      partitionedDataPath,
+      Seq("Date", "RGUID", "Query", "imprs", "clicks"),
+      Some(Seq("Date", "Query")))
+    val partitionedDataDF = spark.read.parquet(partitionedDataPath)
+
+    Seq(nonPartitionedDataPath, partitionedDataPath).foreach { loc =>
+      withSQLConf(
+        IndexConstants.INDEX_LINEAGE_ENABLED -> "true",
+        IndexConstants.REFRESH_DELETE_ENABLED -> "true") {
+        withIndex(indexConfig.indexName) {
+          val dfToIndex =
+            if (loc.equals(nonPartitionedDataPath)) nonPartitionedDataDF else partitionedDataDF
+          hyperspace.createIndex(dfToIndex, indexConfig)
+
+          // delete some source data file.
+          val dataPath =
+            if (loc.equals(nonPartitionedDataPath)) new Path(nonPartitionedDataPath, "*parquet")
+            else new Path(partitionedDataPath + "/Date=2018-09-03/Query=ibraco", "*parquet")
+
+          val dataFileNames = dataPath
+            .getFileSystem(new Configuration)
+            .globStatus(dataPath)
+            .map(_.getPath)
+
+          assert(dataFileNames.length > 0)
+          val fileToDelete = dataFileNames.head
+          FileUtils.delete(fileToDelete)
+
+          val originalIndexDF = spark.read.parquet(s"$systemPath/${indexConfig.indexName}/" +
+            s"${IndexConstants.INDEX_VERSION_DIRECTORY_PREFIX}=0")
+          val delCount =
+            originalIndexDF
+              .filter(s"""${IndexConstants.DATA_FILE_NAME_COLUMN} ==  "$fileToDelete"""")
+              .count()
+
+          hyperspace.refreshIndex(indexConfig.indexName)
```

Comment:
Could you add a E2E test if the refreshed index is applied to a query plan properly?

Reply:
New test cases added under `E2EHyperspaceRulesTests.scala`